internal/webhook-updater/app.go

```diff
@@ -0,0 +1,257 @@
+	}
+
+	quitSignal := registerSignalHandlers()
+	err = app.processFileChanges(client, quitSignal)
+	if err != nil {
+		Log.Errorf("Error watching for certificates: %s", err)
+		os.Exit(1)
+	}
+}
+
+func (app *App) parseAndValidateFlags() error {
+	flag.StringVarP(&app.kubeconfig, "kubeconfig", "c", "", "absolute path to the kuebconfig file")
+	flag.StringVarP(&app.webhookName, "webhook", "w", "", "name of the webhook")
+	flag.StringVarP(&app.caDirectory, "dir", "d", "", "path to the directory that will be watched")
+	flag.StringVarP(&app.caFileName, "file", "f", "", "name of the certificate file in the watched directory")
+	flag.Parse()
+
+	if app.webhookName == "" {
+		return errors.New("webhook parameter cannot be empty")
+	}
+	if !isValidK8sResourceName(app.webhookName) {
+		return errors.New("webhook parameter must be a valid Kubernetes resource name")
+	}
+
+	dirStat, err := os.Stat(app.caDirectory)
+	if err != nil {
+		return err
+	}
+	if !dirStat.IsDir() {
+		return fmt.Errorf("'%s' has to be a directory", app.caDirectory)
+	}
+
+	if app.caFileName == "" {
+		return errors.New("file parameter has to be specified")
+	}
+
+	if app.kubeconfig != "" {
+		stat, err := os.Stat(app.kubeconfig)
+		if err != nil {
+			return err
+		}
+		if stat.IsDir() {
+			return errors.New("kubeconfig does not point to a file")
+		}
+	}
+
+	return nil
+}
+
+func isValidK8sResourceName(name string) bool {
+	nameErrors := validation.NameIsDNSSubdomain(name, false)
+	return len(nameErrors) == 0
+}
+
+func registerSignalHandlers() <-chan struct{} {
+	quit := make(chan struct{})
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, syscall.SIGINT, syscall.SIGTERM)
+	go func() {
+		<-c
+		close(quit)
+	}()
+
+	return quit
+}
+
+func (app *App) processFileChanges(client ClientProxy, quitSignal <-chan struct{}) error {
+	fileChanged, closer, err := watchFsChanges(app.caDirectory)
+	if err != nil {
+		return err
+	}
+	defer closer.Close()
+
+	notify(fileChanged)
+	for {
+		select {
+		case _, ok := <-fileChanged:
+			if !ok {
+				return nil
+			}
+
+			err := app.refreshWebhook(client)
+			if err != nil {
+				// If the webhook does not exist, exit
+				if err == WebhookNotFoundError {
+					return err
+				}
+
+				go func() {
+					time.Sleep(retryInterval)
+					notify(fileChanged)
+				}()
+			}
+		case <-quitSignal:
+			return nil
+		}
+	}
+}
+
+func watchFsChanges(directory string) (fileChanged chan struct{}, closer io.Closer, err error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		Log.Errorf("Failed to create an inotify watcher. Reason: %s", err)
+		return nil, nil, err
+	}
+
+	err = watcher.Add(directory)
+	if err != nil {
+		watcher.Close()
+		Log.Errorf("Failed to establish a watch on %s. Reason: %s", directory, err)
+		return nil, nil, err
+	}
+
+	filesChanged := make(chan struct{}, 1)
+	go func() {
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok {
+					return
+				}
+				if event.Op != fsnotify.Chmod {
```

Comment:
In an event of `fsnotify.Remove`, the process would continuously fail until a new cert file is `fsnotify.Create`d.
Wouldn't it be a good idea to also ignore the `fsnotify.Remove` event and handle the update in the `fsnotify.Create` event that might follow? Because the `caBundle` would not be updated because of `certBytes, err := ioutil.ReadFile(fileName)` in `refreshWebhook`

Reply:
Done